### PR TITLE
Improve battery voltage sag compensation / power supply impedance calculation

### PR DESCRIPTION
**Pull request by shellixyz**

### CODE REVIEW

File: src/main/sensors/battery.c [hunk 1]
```diff
@@ -355,40 +358,67 @@ void powerMeterUpdate(timeUs_t timeDelta)
...
 }
 
-void sagCompensatedVBatUpdate(timeUs_t currentTime)
+void sagCompensatedVBatUpdate(timeUs_t currentTime, timeUs_t timeDelta)
 {
     static timeUs_t recordTimestamp = 0;
     static int32_t amperageRecord;
     static uint16_t vbatRecord;
+    static uint8_t impedanceSampleCount = 0;
+    static pt1Filter_t impedanceFilterState = PT1_FILTER_INITIALIZER;
```
Comment (digitalentity) [inline src/main/sensors/battery.c]: Initializing filter with RC = 0 doesn't make much sense. I don't see where in the code filter is properly initialized with cutoff frequency.

File: src/main/sensors/battery.c [hunk 2]
```diff
@@ -355,40 +358,67 @@ void powerMeterUpdate(timeUs_t timeDelta)
...
 
+            if (impedanceFilterState.state) {
+                impedanceFilterState.RC = impedanceSampleCount > IMPEDANCE_STABLE_SAMPLE_COUNT_THRESH ? 1.2 : 0.5;
+                pt1FilterApply3(&impedanceFilterState, impedanceSample, timeDelta * 1e-6f);
+            } else {
+                pt1FilterReset(&impedanceFilterState, impedanceSample);
+            }
+
+            if (impedanceSampleCount > IMPEDANCE_STABLE_SAMPLE_COUNT_THRESH) {
+                powerSupplyImpedance = lrintf(impedanceFilterState.state);
```
Comment (digitalentity) [inline src/main/sensors/battery.c]: This should be made more generic. Accessing filter internal structure is unreasonable and inherently unsafe. Please consider adding a `pr1FilterGetLastOutput` API or similar.

**Review by digitalentity — CHANGES_REQUESTED** [review]

**Comment by digitalentity** [comment]
Maybe we can also expose calculated battery impedance as an OSD value or log it in a slow blackbox frame?

File: src/main/sensors/battery.c (hunk 1 above)
Comment (shellixyz) [reply src/main/sensors/battery.c]: It is done line 401
`sagCompVBatFilterState.RC = sagCompensatedVBatSample < sagCompVBatFilterState.state ? 40 : 500;`

I will add functions to access the time constant and last value.

**Comment by shellixyz** [comment]
Sure why not.

**Comment by shellixyz** [comment]
I made the requested changes. How would you like the impedance to be formatted on the OSD ? Should we add an ohm symbol ?

**Comment by shellixyz** [comment]
Please merge #2863 to avoid conflicts when adding OSD items

**Review by digitalentity — APPROVED** [review]

**Pull request merged**